backends/p4tools/common/lib/gen_eq.h

```diff
@@ -5,27 +5,23 @@
 
 namespace P4Tools {
 
-/// Generates an equality on a target expression and a keyset expression, recursing into lists.
+/// Generates an semantic equality on two input expressions, recursing into lists and structs.
 /// This supports fuzzy matching on singleton lists: singleton lists are considered the same as
 /// their singleton elements. This is implemented by eagerly recursing into singleton lists before
 /// attempting to generate the equality.
```

Comment:
Please mention a precondition that the types need to be compatible.

Also, I suggest you move the documentation to the function and better yet that you make `equate` into a freestanding function and remove the class altogether (moving the helpers into .cpp).